src,lib: reducing C++ calls of esm legacy main resolve

H4ad opened the pull request:
This PR is related to https://github.com/nodejs/performance/issues/73

## Legacy Main Resolve

@anonrig raised that we could benefit from the implementation of legacyMainResolve on the C++ side, so here's the result so far:

```
                                                                                                                                                                confidence improvement accuracy (*)
esm/esm-legacyMainResolve.js resolvedFile='node_modules/non-exist' packageConfigMain='./index.js' packageJsonUrl='node_modules/test/package.json' n=10000              ***     34.86 %       ±1.41%
esm/esm-legacyMainResolve.js resolvedFile='node_modules/non-exist' packageConfigMain='' packageJsonUrl='node_modules/test/package.json' n=10000                        ***     34.92 %       ±1.52%
esm/esm-legacyMainResolve.js resolvedFile='node_modules/test/index.js' packageConfigMain='./index.js' packageJsonUrl='node_modules/test/package.json' n=10000          ***     34.53 %       ±1.22%
esm/esm-legacyMainResolve.js resolvedFile='node_modules/test/index.js' packageConfigMain='' packageJsonUrl='node_modules/test/package.json' n=10000                    ***     35.39 %       ±1.57%
esm/esm-legacyMainResolve.js resolvedFile='node_modules/test/index.json' packageConfigMain='./index.js' packageJsonUrl='node_modules/test/package.json' n=10000        ***     34.78 %       ±1.30%
esm/esm-legacyMainResolve.js resolvedFile='node_modules/test/index.json' packageConfigMain='' packageJsonUrl='node_modules/test/package.json' n=10000                  ***     35.39 %       ±1.15%
esm/esm-legacyMainResolve.js resolvedFile='node_modules/test/index.node' packageConfigMain='./index.js' packageJsonUrl='node_modules/test/package.json' n=10000        ***     34.75 %       ±1.43%
esm/esm-legacyMainResolve.js resolvedFile='node_modules/test/index.node' packageConfigMain='' packageJsonUrl='node_modules/test/package.json' n=10000                  ***     35.84 %       ±1.33%
                                                                                                                                                                  (**)  (***)
esm/esm-legacyMainResolve.js resolvedFile='node_modules/non-exist' packageConfigMain='./index.js' packageJsonUrl='node_modules/test/package.json' n=10000       ±1.88% ±2.45%
esm/esm-legacyMainResolve.js resolvedFile='node_modules/non-exist' packageConfigMain='' packageJsonUrl='node_modules/test/package.json' n=10000                 ±2.02% ±2.64%
esm/esm-legacyMainResolve.js resolvedFile='node_modules/test/index.js' packageConfigMain='./index.js' packageJsonUrl='node_modules/test/package.json' n=10000   ±1.62% ±2.12%
esm/esm-legacyMainResolve.js resolvedFile='node_modules/test/index.js' packageConfigMain='' packageJsonUrl='node_modules/test/package.json' n=10000             ±2.09% ±2.74%
esm/esm-legacyMainResolve.js resolvedFile='node_modules/test/index.json' packageConfigMain='./index.js' packageJsonUrl='node_modules/test/package.json' n=10000 ±1.74% ±2.27%
esm/esm-legacyMainResolve.js resolvedFile='node_modules/test/index.json' packageConfigMain='' packageJsonUrl='node_modules/test/package.json' n=10000           ±1.53% ±1.99%
esm/esm-legacyMainResolve.js resolvedFile='node_modules/test/index.node' packageConfigMain='./index.js' packageJsonUrl='node_modules/test/package.json' n=10000 ±1.91% ±2.48%
esm/esm-legacyMainResolve.js resolvedFile='node_modules/test/index.node' packageConfigMain='' packageJsonUrl='node_modules/test/package.json' n=10000           ±1.77% ±2.31%

Be aware that when doing many comparisons the risk of a false-positive
result increases. In this case, there are 8 comparisons, you can thus
expect the following amount of false-positive results:
  0.40 false positives, when considering a   5% risk acceptance (*, **, ***),
  0.08 false positives, when considering a   1% risk acceptance (**, ***),
  0.01 false positives, when considering a 0.1% risk acceptance (***)
```
> To be able to run the benchmark, first build the `old` node using [this branch](https://github.com/h4ad-forks/node/tree/bench/add-esm-legacy-main-resolve).

About the benchmark, I needed to create another benchmark just for this function to be able to see the improvements since the results of [esm-defaultResolver.js](https://github.com/nodejs/node/blob/main/benchmark/esm/esm-loader-defaultResolve.js) was not reflecting the real impact of this change. 

## Potential optimizations

`legacyMainResolve` also calls `fileURLToPath` and didn't have any version on the C++ side, so I needed to rewrite this function.

I haven't benchmarked the JS version yet, but I think we can add some benchmarks and see if we can replace the JS version with the C++ version, I've tried to make the code as easy as possible to support exposing on the JS side in the future.

## Tasks

- [x] Rewrite legacyMainResolve
- [x] Rewrite fileURLToPath
- [x] Add benchmark for legacyMainResolve
- [x] Add tests for legacyMainResolve

## Acknowledgments

Huge thanks to @anonrig for the support and early review of this code and @RafaelGSS for the hints about the benchmark and permission model.

<!--
Before submitting a pull request, please read
https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md.

Commit message formatting guidelines:
https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

RafaelGSS commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,270 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+
+    file_path_agg = &file_path_url.value();
+
+    if (!FileURLToPathImpl(env, file_path_agg, &initial_file_path)) return;
```
I really feel you are leaking memory here. Please, consider passing variables by reference instead of pointers or using a std container for that, such as `std::unique_ptr `

anonrig replied:
Pass by reference 👍

H4ad replied:
I think now is right.

RafaelGSS commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,270 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+    THROW_IF_INSUFFICIENT_PERMISSIONS(
+        env, permission::PermissionScope::kFileSystemRead, file_path);
+
+    if (UNSAFE_FileExist(env, file_path)) return args.GetReturnValue().Set(0);
```
Wouldn't it block the event loop?

RafaelGSS commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,270 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+
+// this method skip the permission model, you MUST verify the permission
+// manually if you want to use this method
+bool UNSAFE_FileExist(Environment* env, const std::string& file_path) {
```
It's way better/cleaner to have the THROW_IF_INSUFFICIENT_PERMISSIONS here.

H4ad replied:
If you think is better, I can rewrite `InternalModuleStat` and reuse this same check for both methods, but I will need to create a signature like this:

```c++
int FileExist(Environment* env, const std::string& file_path) {
```

- 0: Do not exist
- 1: Exist
- -1: Error Permission

And the consumer must check for the result.

anonrig replied:
You can return an enum value to make it more readable

RafaelGSS reviewed (commented):
IMO the new implementation is prone to memory leaks and it's quite complex to read

RafaelGSS reviewed (commented):
IMO the new implementation is prone to memory leaks and it's quite complex to read

H4ad commented on src/node_errors.h:
```diff
@@ -68,11 +68,14 @@ void AppendExceptionLine(Environment* env,
   V(ERR_INVALID_ARG_VALUE, TypeError)                                          \
   V(ERR_OSSL_EVP_INVALID_DIGEST, Error)                                        \
   V(ERR_INVALID_ARG_TYPE, TypeError)                                           \
+  V(ERR_INVALID_FILE_URL_HOST, TypeError)                                      \
```
`docs/api/errors.md`?

anonrig commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -153,44 +160,23 @@ function fileExists(url) {
-      return guess;
+  const resolvedOption = URLLegacyMainResolve(packageJSONUrl.href, packageConfig.main);
+
+  if (resolvedOption < 0) {
```
Can you add comments to each magical number comparison?

anonrig commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -153,44 +160,23 @@ function fileExists(url) {
-  throw new ERR_MODULE_NOT_FOUND(
-    fileURLToPath(new URL('.', packageJSONUrl)), fileURLToPath(base));
+
+  const baseUrl = resolvedOption <= 6 ? `./${packageConfig.main}` : '';
```
Can you add a comment to here?

anonrig commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,270 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
   }
 }
 
+bool FileURLToPathWin32Impl(Environment* env,
```
Please initialize a single FileURLToPath function for all environments. No need to name this function Win32.

anonrig commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,270 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+  return true;
+}
+
+bool FileURLToPathImpl(Environment* env,
```
Having 3 different File url to path function makes it really hard to read/understand.

anonrig commented on src/node_errors.h:
```diff
@@ -163,6 +166,7 @@ ERRORS_WITH_CODE(V)
   V(ERR_INVALID_STATE, "Invalid state")                                        \
   V(ERR_INVALID_THIS, "Value of \"this\" is the wrong type")                   \
   V(ERR_INVALID_TRANSFER_OBJECT, "Found invalid object in transferList")       \
+  V(ERR_INVALID_URL_SCHEME, "The URL must be of scheme file:")                 \
```
the name and the error message creates confusion.

H4ad replied:
I got from: 
https://github.com/nodejs/node/blob/cd6bd3f0fb135d3ef7cc847d62205c6f0d7a47c8/lib/internal/errors.js#L1381-L1390

And:

https://github.com/nodejs/node/blob/cd6bd3f0fb135d3ef7cc847d62205c6f0d7a47c8/lib/internal/url.js#L1359-L1360C29

What I can do is remove this default message and pass the message when I create the error.

anonrig commented on src/node_errors.h:
```diff
@@ -68,11 +68,14 @@ void AppendExceptionLine(Environment* env,
   V(ERR_INVALID_STATE, Error)                                                  \
   V(ERR_INVALID_THIS, TypeError)                                               \
   V(ERR_INVALID_TRANSFER_OBJECT, TypeError)                                    \
+  V(ERR_INVALID_URL_SCHEME, TypeError)                                         \
```
url has protocol. i think this needs to be invalid_url_protocol

H4ad replied:
I just follow the original implementation:

https://github.com/nodejs/node/blob/cd6bd3f0fb135d3ef7cc847d62205c6f0d7a47c8/lib/internal/url.js#L1359-L1360C29

Should I change the error message?

anonrig commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,222 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+
+  if (hostname.size() > 0) {
+    std::string error_message =
+        std::string("File URL host must be \"localhost\" or empty on ") +
```
You didn't check if file url host is localhost.

H4ad replied:
I just follow the original implementation:

https://github.com/nodejs/node/blob/cd6bd3f0fb135d3ef7cc847d62205c6f0d7a47c8/lib/internal/errors.js#L1307-L1308

Should I change it?

anonrig replied:
@nodejs/url any ideas?

aduh95 replied:
`node -p 'url.fileURLToPath("file://localhost/dev/null")'` does return `/dev/null` without any error for me. It should be noted that `out/Release/node -p 'new URL("file://localhost/dev/null").hostname'` returns an empty string, which I suppose is because https://url.spec.whatwg.org/#file-host-state says `If host is "localhost", then set host to the empty string`.

anonrig commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,222 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+
+  int first_percent = 0;
+
+  for (std::string::size_type i = 0; (i + 2) < pathname.size(); ++i) {
```
This should be size_t for readability

anonrig commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,222 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+      first_percent = i;
+    }
+
+    if (pathname[i + 1] != '2' || (pathname[i + 2] | 0x20) != 102) continue;
```
For readability, i think it is better to just have a single if statement for false case.

anonrig commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,222 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+          // the default behavior when do not have permission is to return
+          // and exit the execution of the method as soon as possible
+          // the internal function will throw the exception
+          return;
```
It should never execute in this line. I don't think we need this enum.

H4ad replied:
`FilePathIsFile` now verifies the permission model, I need this case because the user can try to import a library using this old method and that library was protected by the permission model, then it should throw an error.

anonrig commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,222 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+
+    if (!FileURLToPathImpl(env, file_path_agg, initial_file_path)) return;
+
+    FromNamespacedPath(&(*initial_file_path));
```
I recommend dropping pointer and just pass the reference:

```suggestion
    FromNamespacedPath(initial_file_path);
```

RaisinTen replied:
Since FromNamespacedPath() mutates the string, it can't be a const ref, so I think we should still pass a pointer (it's mentioned in the c++ style guide). If this is for readability, what about using get():
```suggestion
    FromNamespacedPath(initial_file_path.get());
```

H4ad replied:
Instead of using `unique_ptr`, I change to `std::string&` on the method and this line became `FromNamespacedPath(&initial_file_path);`, much simpler.

anonrig commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,222 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
 
+static bool FileURLToPathImpl(Environment* env,
+                       ada::url_aggregator* file_url,
+                       std::unique_ptr<std::string>& result_file_path) {
```
I don't understand why we don't do this?

```suggestion
                       std::string& result_file_path) {
```

H4ad replied:
Fixed!

H4ad replied:
But this causes me a lint issue:

```
Running C++ linter...
/home/h4ad/Projects/opensource/node/tools/cpplint.py:57: DeprecationWarning: module 'sre_compile' is deprecated
  import sre_compile
src/node_file.cc:2735:  Is this a non-const reference? If so, make const or use a pointer: std::string& result_file_path  [runtime/references] [2]
Done processing src/node_file.cc
```

If I make it const, I will not be able to modify it, right?

debadree25 commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,229 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+#endif
+}
+
+FilePathIsFileReturnType FilePathIsFile(Environment* env,
```
This function seems to be the same as https://github.com/nodejs/node/blob/4bb06dbd0a87e5b2430780ddceec08015f8c496a/src/node_file.cc#L1124

maybe we could DRY the implementation of both here? (this not really blocking just a observation)

H4ad replied:
Initially, I was thinking of doing that but I cannot join both functions since I don't know how I can represent the state of the `THROW_IF_INSUFFICIENT_PERMISSIONS`, when the user doesn't have permissions, we should stop the execution of the function, so I need to return not only an `int` but also another state to tell the caller that the user doesn't have permission.

Because of this, I prefer to return an enum for `FilePathIsFile` and not over complicate both implementations.

anonrig replied:
@RafaelGSS might help

aduh95 commented:
> `legacyMainResolve` also calls `fileURLToPath` and didn't have any version on the C++ side, so I needed to rewrite this function.

Maybe we should first test/benchmark and land the `fileURLToPath` rewrite, which would simplify this PR. wdyt?

H4ad commented:
@aduh95 Here's the benchmark for rewrite of `fileURLToPath`:

```
                                                                                         confidence improvement accuracy (*)    (**)   (***)
url/file-url-to-path.js n=100000 url='(url) file:///home/user/test/index.js'                    ***    -95.88 %       ±3.25%  ±4.44%  ±6.03%
url/file-url-to-path.js n=100000 url='(url) file:///home/user/test%20index.js'                  ***    -95.19 %       ±2.98%  ±4.07%  ±5.52%
url/file-url-to-path.js n=100000 url='(url) file:///home/user/test%2Findex.js'                  ***      8.66 %       ±2.71%  ±3.67%  ±4.90%
url/file-url-to-path.js n=100000 url='(url) file://google.com/home/user/test%2Findex.js'                 0.68 %       ±2.75%  ±3.72%  ±4.99%
url/file-url-to-path.js n=100000 url='(url) http://google.com/home/test.js'                     ***    -11.02 %       ±2.27%  ±3.05%  ±4.03%
url/file-url-to-path.js n=100000 url='file:///home/user/test/index.js'                          ***     97.59 %       ±7.74% ±10.45% ±13.92%
url/file-url-to-path.js n=100000 url='file:///home/user/test%20index.js'                        ***     86.08 %       ±7.51% ±10.15% ±13.52%
url/file-url-to-path.js n=100000 url='file:///home/user/test%2Findex.js'                        ***    195.47 %       ±3.95%  ±5.35%  ±7.14%
url/file-url-to-path.js n=100000 url='file://google.com/home/user/test%2Findex.js'              ***    166.80 %       ±3.19%  ±4.29%  ±5.66%
url/file-url-to-path.js n=100000 url='http://google.com/home/test.js'                           ***    145.18 %       ±4.26%  ±5.76%  ±7.68%
url/file-url-to-path.js n=100000 url='not-even-a-url'                                           ***    168.09 %       ±7.25%  ±9.87% ±13.32%

Be aware that when doing many comparisons the risk of a false-positive
result increases. In this case, there are 11 comparisons, you can thus
expect the following amount of false-positive results:
  0.55 false positives, when considering a   5% risk acceptance (*, **, ***),
  0.11 false positives, when considering a   1% risk acceptance (**, ***),
  0.01 false positives, when considering a 0.1% risk acceptance (***)
```

Branch: https://github.com/nodejs/node/compare/main...h4ad-forks:node:perf/file-url-to-path

aduh95 commented:
@H4ad I meant to have it as a separate PR.

RafaelGSS commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -153,44 +160,29 @@ function fileExists(url) {
-      return guess;
+  const resolvedOption = URLLegacyMainResolve(packageJSONUrl.href, packageConfig.main);
+
+  // less than zero means that happened some error
```
this is implicit 
```suggestion
```

RafaelGSS commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -153,44 +160,29 @@ function fileExists(url) {
+  const resolvedOption = URLLegacyMainResolve(packageJSONUrl.href, packageConfig.main);
+
+  // less than zero means that happened some error
+  // -2: LegacyMainResolveReturnType::kInvalidUrl
```
maybe store it in a variable so you can compare `resolvedOption === invalidUrlReturn`

RafaelGSS commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -153,44 +160,29 @@ function fileExists(url) {
+  // -1: LegacyMainResolveReturnType::kModuleNotFound
+  if (resolvedOption < 0) {
+    if (resolvedOption === -2) {
+      throw new ERR_INVALID_URL(packageConfig.main || './index.js');
```
Is it a new exception for this path? I don't see it in the previous version.

More specifically, are you sure it's not a breaking change?

H4ad replied:
It's not a new exception, this if statement represents this line:

https://github.com/nodejs/node/blob/5c27cc2afbe77d4978a7c773c1dcf2e738aad7e7/lib/internal/modules/esm/resolve.js#L159-L160

There are two cases that the URL can be invalid, when `packageConfig.main` represents an invalid URL or when the `packageJSONUrl` cannot be parsed as the base URL.

anonrig replied:
It's not a breaking change. `new URL` can throw `ERR_INVALID_URL`

RafaelGSS commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -153,44 +160,29 @@ function fileExists(url) {
+  // 0-6: when packageConfig.main is defined
+  // 7-9: when packageConfig.main is NOT defined,
+  //      or when the previous case didn't found the file
+  const baseUrl = resolvedOption <= 6 ? `./${packageConfig.main}` : '';
```
I'd compare it with a variable instead, just to make it more readable. Something like:

`isPackageConfigMainDefined(resolvedOption)`

RafaelGSS commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -1078,6 +1070,7 @@ module.exports = {
   packageExportsResolve,
   packageImportsResolve,
   throwIfInvalidParentURL,
+  legacyMainResolve,
```
you are exporting it just for benchmark?

H4ad replied:
Yes

RafaelGSS commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,219 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+#if _WIN_32
+  std::string pathname_escaped_slash;
+
+  for (size_t i = 0; (i + 2) < pathname.size(); i++) {
```
Can you include a few comments in this section? I also suggest removing the blank lines.

H4ad commented:
@aduh95 Yeah, my idea was to show you if you really want to do that with the performance benefits from it.

To be honest, in this form, without refactoring the whole ESM to avoid `new URL` and manipulate the data directly as a string, we will not see a performance improvement, it will be the complete opposite.

I think is better to go in the opposite direction, merge this one and then evaluate the benefits of exposing `fileURLToPath`, because as I said, will probably lead to a refactoring not only in the function itself but also in the places where they call to take the fastest path, which is manipulating directly the string.

GeoffreyBooth commented:
> url/file-url-to-path.js n=100000 url=’(url) file:///home/user/test/index.js’                    ***    -95.88 %       ±3.25%  ±4.44%  ±6.03%

This seems to show that in the most common case of a `file://` URL, this is 96% slower?

H4ad commented:
@GeoffreyBooth Exactly, the way the `fileURLToPath` is used today, simply don't worth exposing the C++ version.

GeoffreyBooth commented:
> @GeoffreyBooth Exactly, the way the `fileURLToPath` is used today, simply don’t worth exposing the C++ version.

Okay, so then should we close this PR? Is there something else of value being proposed here?

H4ad commented:
This PR is not about `fileURLToPath`, is about `LegacyMainResolve`.

The improvements for `LegacyMainResolve` is about 60~70%.

The rewrite of `fileURLToPath` was only because it was needed by `LegacyMainResolve`. Aduh suggested that maybe we could create a PR for `fileURLToPath` but as we've seen, it's not worth it, that's it.

H4ad commented:
The CI for Test Asan is failing due some memory issue:

```
==88318==ERROR: AddressSanitizer: stack-use-after-scope on address 0x7ffd9ad085f8 at pc 0x00000431f995 bp 0x7ffd9ad07570 sp 0x7ffd9ad07568
READ of size 16 at 0x7ffd9ad085f8 thread T0
    #0 0x431f994 in ada::url_aggregator ada::parser::parse_url<ada::url_aggregator>(std::basic_string_view<char, std::char_traits<char> >, ada::url_aggregator const*) (/home/runner/work/node/node/out/Release/node+0x431f994)
    #1 0x4312e13 in tl::expected<ada::url_aggregator, ada::errors> ada::parse<ada::url_aggregator>(std::basic_string_view<char, std::char_traits<char> >, ada::url_aggregator const*) (/home/runner/work/node/node/out/Release/node+0x4312e13)
    #2 0xfc3b97 in node::fs::LegacyMainResolve(v8::FunctionCallbackInfo<v8::Value> const&) (/home/runner/work/node/node/out/Release/node+0xfc3b97)
    #3 0x1779bab in v8::internal::MaybeHandle<v8::internal::Object> v8::internal::(anonymous namespace)::HandleApiCallHelper<false>(v8::internal::Isolate*, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::FunctionTemplateInfo>, v8::internal::Handle<v8::internal::Object>, unsigned long*, int) (/home/runner/work/node/node/out/Release/node+0x1779bab)
    #4 0x17779b4 in v8::internal::Builtin_HandleApiCall(int, unsigned long*, v8::internal::Isolate*) (/home/runner/work/node/node/out/Release/node+0x17779b4)
    #5 0x3402e35 in Builtins_CEntry_Return1_ArgvOnStack_BuiltinExit (/home/runner/work/node/node/out/Release/node+0x3402e35)

Address 0x7ffd9ad085f8 is located in stack of thread T0 at offset 1304 in frame
    #0 0xfc364f in node::fs::LegacyMainResolve(v8::FunctionCallbackInfo<v8::Value> const&) (/home/runner/work/node/node/out/Release/node+0xfc364f)
```

> Ref: https://github.com/nodejs/node/actions/runs/5189308542/jobs/9354128411?pr=48325

Does anyone know what could be the cause?

I tried to reproduce it locally but I couldn't, so I have no idea how to fix it.

RaisinTen commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,222 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+
+  Environment* env = Environment::GetCurrent(args);
+
+  std::string_view base_path =
+      Utf8Value(env->isolate(), args[0].As<String>()).ToStringView();
+  auto base = ada::parse<ada::url_aggregator>(base_path);
```
```suggestion
  Utf8Value base_path(env->isolate(), args[0].As<String>());
  auto base = ada::parse<ada::url_aggregator>(base_path.ToStringView());
```
The Utf8Value object needs to be on the stack for the returned string view to be valid as the actual string is owned by the Utf8Value object. This should fix the ASAN error you're facing.

H4ad replied:
It worked, thanks!

anonrig commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,238 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+        std::string("./") + package_config_main, &base.value());
+
+    if (!file_path_url) {
+      return args.GetReturnValue().Set(
```
Can you throw an error on here as well?

anonrig commented on src/node_file.h:
```diff
@@ -55,6 +55,35 @@ enum class FsStatFsOffset {
+// 0-6: when packageConfig.main is defined
+// 7-9: when packageConfig.main is NOT defined,
+//      or when the previous case didn't found the file
+const std::vector<std::string> legacy_main_extensions = {"",
```
This can be a `std::array` since this variable is not mutable. Than we can convert this to `constexpr std::array`

anonrig commented on src/node_file.h:
```diff
@@ -55,6 +55,35 @@ enum class FsStatFsOffset {
+                                                         ".node"};
+// define the final index of the algorithm resolution
+// when packageConfig.main is defined.
+const uint8_t legacy_main_extensions_with_main_end = 7;
```
Can you make this private in bindingdata? no need to expose these variables to all files that includes the header file.

H4ad replied:
If I do that, won't I need to put the `LegacyMainResolve` method inside the BindingData to be able to access that private variable?

Also, by doing this, why not put `legacy_main_resolve` and the other constants too?

anonrig replied:
Yes. I think we should.

anonrig commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,238 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+
+static void LegacyMainResolve(const FunctionCallbackInfo<Value>& args) {
+  CHECK_GE(args.Length(), 1);
+  // packageJSONUrl
```
```suggestion
```

anonrig commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,238 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+
+  Environment* env = Environment::GetCurrent(args);
+
+  Utf8Value utf8_base_path = Utf8Value(env->isolate(), args[0].As<String>());
```
```suggestion
  Utf8Value utf8_base_path(env->isolate(), args[0].As<String>());
```

anonrig commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -132,13 +134,36 @@ function emitLegacyIndexDeprecation(url, packageJSONUrl, base, main) {
+];
+
+const mainResolveReturnType = {
+  kInvalidUrl: -2,
+  kModuleNotFound: -1,
```
I'm in favor of removing these error cases, and handling (throwing) them on C++ side.

anonrig commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,238 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
   }
 }
 
+static bool FileURLToPathImpl(
```
```suggestion
static bool FileURLToPath(
```

anonrig commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,238 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+    return args.GetReturnValue().Set(LegacyMainResolveReturnType::kInvalidUrl);
+  }
+
+  if (!FileURLToPathImpl(env, file_path_url.value(), initial_file_path)) return;
```
```suggestion
  if (!FileURLToPath(env, file_path_url.value(), initial_file_path)) return;
```

H4ad replied:
The idea of using `Impl` in the end was because of that method could be used to implement the real `fileURLToPath` on the JS side.

But after the benchmarks, probably doesn't worth having this method exported.

anonrig commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,238 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+  }
+
+  return args.GetReturnValue().Set(
+      LegacyMainResolveReturnType::kModuleNotFound);
```
Can you throw an error in here, instead of returning `kModuleNotFound`?

H4ad replied:
Throw error on C++ side is faster than JS side?

anonrig replied:
It's not faster, but we don't need to make it faster, since this will stop the execution. So the question is `Do we need to crash faster`?

anonrig commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,238 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+#endif
+}
+
+FilePathIsFileReturnType FilePathIsFile(Environment* env,
```
Since we are not exposing this to somewhere else, can you make it private and move it inside BindingData class?

anonrig reviewed (commented):
I'm OK with the implementation. I've added some comments regarding how error states are handled between C++ and JS. Removing those from JS would make it much more readable imho.

H4ad commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,290 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+    if (!FileURLToPath(env, base_url.value(), module_base)) return;
+  } else {
+    std::string received = args.Length() >= 3
+                               ? "How to Serialize v8::Value to a type?"
```
@anonrig Do you know how I can get the type of v8::Value easily here?

I need to throw this error to keep the behavior consistent, this error is from `fileURLToPath` and it can happen when `base` is undefined or other value.

H4ad replied:
I talked with Rafael, his suggestion was just to remove the "Receive " part since we cannot get easily the V8::Value type.

anonrig commented on src/node_file.cc:
```diff
@@ -2826,6 +3109,19 @@ InternalFieldInfoBase* BindingData::Serialize(int index) {
+                                             Local<ObjectTemplate> target) {
+  Isolate* isolate = isolate_data->isolate();
+
+  SetMethod(
```
```suggestion
  SetMethodNoSideEffect(
```

H4ad replied:
Calling `uv` is not considered as side effect?

anonrig replied:
You're right.

anonrig commented on src/node_file.h:
```diff
@@ -79,12 +86,30 @@ class BindingData : public SnapshotableObject {
+  static FilePathIsFileReturnType FilePathIsFile(Environment* env,
+                                                 const std::string& file_path);
+
+  static const std::array<std::string, 10> legacy_main_extensions;
```
can you move the std::array contents to here, instead of defining them in .cc?

H4ad replied:
Do not compile:

```
../../src/node_file.h:110:44: error: in-class initialization of static data member ‘const std::array<std::__cxx11::basic_string<char>, 10> node::fs::BindingData::legacy_main_extensions’ of non-literal type
  110 |   static const std::array<std::string, 10> legacy_main_extensions = {
      |                                            ^~~~~~~~~~~~~~~~~~~~~~
../../src/node_file.h:120:12: error: temporary of non-literal type ‘std::__cxx11::basic_string<char>’ in a constant expression
  120 |     ".node"};
```

Maybe it could compile if we were using `constexpr`, but since it didn't work with `std::string` :/

anonrig replied:
This article can help: https://stackoverflow.com/questions/71882752/c-constexpr-stdarray-of-string-literals

anonrig replied:
Yoo could use std::string_view instead of string as well.

H4ad replied:
I use this array of strings to concat with the file path, it should be string since using string_view will lead me to do `std::string(legacy_main_extensions[i])`, which I don't think is a good behavior.

What do you think?

anonrig replied:
yeah you are right. lets at least move the declaration to header file.

H4ad replied:
To be able to do that, I need to move the declaration of `legacy_main_extensions` outside of `BindingData`, it's okay?

anonrig commented on src/node_file.h:
```diff
@@ -79,12 +86,30 @@ class BindingData : public SnapshotableObject {
+  static const std::array<std::string, 10> legacy_main_extensions;
+  // define the final index of the algorithm resolution
+  // when packageConfig.main is defined.
+  static const uint8_t legacy_main_extensions_with_main_end = 7;
```
```suggestion
  constexpr uint8_t legacy_main_extensions_with_main_end = 7;
```

anonrig commented on src/node_file.h:
```diff
@@ -79,12 +86,30 @@ class BindingData : public SnapshotableObject {
+  static const uint8_t legacy_main_extensions_with_main_end = 7;
+  // define the final index of the algorithm resolution
+  // when packageConfig.main is NOT defined
+  static const uint8_t legacy_main_extensions_package_fallback_end = 10;
```
```suggestion
  constexpr uint8_t legacy_main_extensions_package_fallback_end = 10;
```

anonrig commented on src/node_file.h:
```diff
@@ -79,12 +86,30 @@ class BindingData : public SnapshotableObject {
+  static FilePathIsFileReturnType FilePathIsFile(Environment* env,
+                                                 const std::string& file_path);
+
+  static const std::array<std::string, 10> legacy_main_extensions;
```
```suggestion
  constexpr std::array<std::string, 10> legacy_main_extensions = {...};
```

H4ad replied:
Do not compile: https://stackoverflow.com/a/71882832

```
In file included from ../../src/dataqueue/queue.h:9,
                 from ../../test/cctest/test_dataqueue.cc:1:
../../src/node_file.h:110:3: error: non-static data member ‘legacy_main_extensions’ declared ‘constexpr’
  110 |   constexpr std::array<std::string, 10> legacy_main_extensions = {
      |   ^~~~~~~~~
[4/140] CXX obj/src/dataqueue/libnode.queue.o
```

anonrig reviewed (approved):
I've added some small comments. After they're resolved, I think this PR is ready to get reviewed & merged.

aduh95 commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -153,44 +175,15 @@ function fileExists(url) {
-  // Not found.
-  throw new ERR_MODULE_NOT_FOUND(
-    fileURLToPath(new URL('.', packageJSONUrl)), fileURLToPath(base));
+  const baseStringified = isURL(base) ? base.href : base;
```
Isn't that more performant?
```suggestion
  const baseStringified = `${base}`;
```

H4ad replied:
But `base` can be `undefined`, so it will be stringified as `'undefined'` and this will not lead to the same error as the previous version.

GeoffreyBooth replied:
I think this matches the previous version and is perhaps slightly faster?

```suggestion
  const baseStringified = base?.href;
```

H4ad replied:
Accords the method signature, `base` can be a string too, the alternative could be `base?.href ?? base`, but I think checking using `isURL` will be faster and more reliable.

GeoffreyBooth replied:
Could we simplify the method signature so that `base` is always passed in as either a string or as a URL, rather than accepting either? I would think that eliminating the `isURL` call should make this slightly faster.

H4ad replied:
`base` is used as a URL in many other places, converting it to a string will not be good for performance, so I will keep it as is.

aduh95 commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -132,13 +133,34 @@ function emitLegacyIndexDeprecation(url, packageJSONUrl, base, main) {
+  './index.node',
+];
+
+const mainResolveReturnType = {
```
```suggestion
const legacyMainResolveReturnType = {
```

H4ad commented:
CI is failing https://ci.nodejs.org/job/node-test-binary-armv7l/RUN_SUBSET=js,nodes=ubuntu2004-armv7l/6181/consoleFull but with an error that I think is not related with this PR:

```
not ok 3740 sequential/test-perf-hooks
01:20:33   ---
01:20:33   duration_ms: 618.48400
01:20:33   severity: fail
01:20:33   exitcode: 1
01:20:33   stack: |-
01:20:33     {
01:20:33       name: 'node',
01:20:33       entryType: 'node',
01:20:33       startTime: 0,
01:20:33       duration: { around: 404.7198510169983 },
01:20:33       nodeStart: { around: 0 },
01:20:33       v8Start: { around: 0 },
01:20:33       bootstrapComplete: { around: 404.6602909564972, delay: 2500 },
01:20:33       environment: { around: 0 },
01:20:33       loopStart: -1,
01:20:33       loopExit: -1
01:20:33     }
01:20:33     node:assert:399
01:20:33         throw err;
01:20:33         ^
01:20:33     
01:20:33     AssertionError [ERR_ASSERTION]: environment: 250.5410280227661 >= 250
01:20:33         at checkNodeTiming (/home/iojs/build/workspace/node-test-binary-armv7l/test/sequential/test-perf-hooks.js:31:7)
01:20:33         at Object.<anonymous> (/home/iojs/build/workspace/node-test-binary-armv7l/test/sequential/test-perf-hooks.js:43:1)
01:20:33         at Module._compile (node:internal/modules/cjs/loader:1255:14)
01:20:33         at Module._extensions..js (node:internal/modules/cjs/loader:1309:10)
01:20:33         at Module.load (node:internal/modules/cjs/loader:1113:32)
01:20:33         at Module._load (node:internal/modules/cjs/loader:960:12)
01:20:33         at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:83:12)
01:20:33         at node:internal/main/run_main_module:23:47 {
01:20:33       generatedMessage: false,
01:20:33       code: 'ERR_ASSERTION',
01:20:33       actual: false,
01:20:33       expected: true,
01:20:33       operator: '=='
01:20:33     }
01:20:33     
01:20:33     Node.js v21.0.0-pre
```

Any hints about what I should do?

H4ad commented:
Also, I updated the benchmark, with the latest changes, the improvement is almost 200%.

But one thing that is making me a little skeptical about this number is this benchmark:

```
                                                                               confidence improvement accuracy (*)   (**)   (***)
esm/esm-loader-defaultResolve.js specifier='./relative-existing.js' n=10000                    4.83 %       ±6.85% ±9.20% ±12.15%
esm/esm-loader-defaultResolve.js specifier='./relative-nonexistent.js' n=10000                 1.95 %       ±4.28% ±5.74%  ±7.56%
esm/esm-loader-defaultResolve.js specifier='node:os' n=10000                                  -1.98 %       ±5.33% ±7.17%  ±9.47%
esm/esm-loader-defaultResolve.js specifier='node:prefixed-nonexistent' n=10000                -3.48 %       ±5.01% ±6.71%  ±8.83%
esm/esm-loader-defaultResolve.js specifier='unprefixed-existing' n=10000                      -0.32 %       ±4.83% ±6.47%  ±8.51%
esm/esm-loader-defaultResolve.js specifier='unprefixed-nonexistent' n=10000                    0.16 %       ±1.99% ±2.67%  ±3.52%

Be aware that when doing many comparisons the risk of a false-positive
result increases. In this case, there are 6 comparisons, you can thus
expect the following amount of false-positive results:
  0.30 false positives, when considering a   5% risk acceptance (*, **, ***),
  0.06 false positives, when considering a   1% risk acceptance (**, ***),
  0.01 false positives, when considering a 0.1% risk acceptance (***)
```

Only `unprefixed-existing` calls `legacyMainResolve`, and we don't see the performance reflected in this benchmark.

I wrote a little benchmark and we came from `36k` op/s to `63k` op/s, which is a real improvement, so does it worth introducing 400+ lines for this amount of improvement?

GeoffreyBooth commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -153,44 +175,15 @@ function fileExists(url) {
+  const resolvedOption = URLLegacyMainResolve(packageJSONUrl.href, packageConfig.main, baseStringified);
+
+  const baseUrl = resolvedOption <= legacyMainResolveReturnType.kResolvedByMainIndexNode ? `./${packageConfig.main}` : '';
+  const guess = new URL(baseUrl + legacyMainResolveExtensions[resolvedOption], packageJSONUrl);
```
It’s not really a guess at this point, is it? We know that this is the resolution?

H4ad replied:
It's a bad name, yeah, is a real resolution determined by URLLegacyMainResolve.

GeoffreyBooth commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -132,13 +133,34 @@ function emitLegacyIndexDeprecation(url, packageJSONUrl, base, main) {
+  './index.node',
+];
+
+const legacyMainResolveReturnType = {
+  // 0-6: when packageConfig.main is defined
+  kResolvedByMain: 0,
+  kResolvedByMainJs: 1,
+  kResolvedByMainJson: 2,
+  kResolvedByMainNode: 3,
+  kResolvedByMainIndexJs: 4,
+  kResolvedByMainIndexJson: 5,
+  kResolvedByMainIndexNode: 6,
+  // 7-9: when packageConfig.main is NOT defined,
+  //      or when the previous case didn't found the file
+  kResolvedByPackageAndJs: 7,
+  kResolvedByPackageAndJson: 8,
+  kResolvedByPackageAndNode: 9,
+};
```
It took me a while to realize that these values are the indexes of the `legacyMainResolveExtensions` array. Is it faster to do it this way than a merged data structure like an object or `SafeMap`?

If this is the fastest option, perhaps name this object `legacyMainResolveExtensionsIndexes` to make the connection clearer?

GeoffreyBooth commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -36,8 +36,9 @@ const preserveSymlinksMain = getOptionValue('--preserve-symlinks-main');
-const { URL, pathToFileURL, fileURLToPath, isURL, toPathIfFileURL } = require('internal/url');
+const { URL, pathToFileURL, fileURLToPath, isURL } = require('internal/url');
 const { canParse: URLCanParse } = internalBinding('url');
+const { legacyMainResolve: URLLegacyMainResolve } = internalBinding('fs');
```
```suggestion
const { legacyMainResolve: FSLegacyMainResolve } = internalBinding('fs');
```

GeoffreyBooth reviewed (commented):
I can’t review the C++, but the JavaScript mostly looks great. Thanks for doing this!

Is this only to improve performance of parsing `"main"` from `package.json`? I assume that we don’t need similar efforts for `"exports"` because those keys are already relative URLs and they’re already explicit?

GeoffreyBooth reviewed (approved):
The JavaScript looks good to me, if @anonrig vouches for the C++.

aduh95 commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,287 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+  result_file_path = decoded_pathname.substr(1);
+
+  return true;
+#else
```
```suggestion
#else  // _WIN32
```

aduh95 commented on src/node_file.cc:
```diff
@@ -2727,6 +2730,287 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+  result_file_path = ada::unicode::percent_decode(pathname, first_percent);
+
+  return true;
+#endif
```
```suggestion
#endif  // _WIN32
```

H4ad commented:
I added some tests to ensure the correct resolution of `legacyMainResolve`, this also helped me catch some edge cases related to the errors.

I just added some good paths and check about the errors, and I didn't add any test to ensure the order of the resolution (if has index.json and index.js, should return index.js first). If you think is necessary, I can add those tests too.

RafaelGSS reviewed (commented):
Can you please include a test to reach the kThrowInsufficientPermissions?

H4ad commented on test/es-module/test-cjs-legacyMainResolve-permission.js:
```diff
@@ -0,0 +1,108 @@
+            );
+
+            assert.throws(() => legacyMainResolve(packageJsonUrl, packageConfig, base), {
+              code: 'ERR_ACCESS_DENIED',
```
I could not make the test run with the permission model, I sent you a message on Slack with the details since I don't know if it is a bug on the permission model.

H4ad replied:
It was indeed a bug, a PR was created to fix it, waiting for https://github.com/nodejs/node/pull/48531 to get back to work on testing for the permission model.

anonrig commented:
@H4ad it’s really hard to keep track of what you’ve changed when you force push on a single commit everytime, and makes this PR hard to merge. I recommend using git commit fixup or just add a new commit for every fix.

H4ad commented:
@anonrig Sorry, I thought it was better to just keep a single commit, that's why I always push force, but I'll be using fixup for future fixes.

RafaelGSS reviewed (commented):
@H4ad  Can you update the benchmark according to the updates?

anonrig commented:
Benchmark CI: https://ci.nodejs.org/view/Node.js%20benchmark/job/benchmark-node-micro-benchmarks/1342/

ruyadorno commented:
This commit does not land cleanly on `v18.x-staging` and will need manual backport in case we want it in **v18**.

H4ad commented:
@ruyadorno Hey, can I help with something?

RaisinTen commented:
@H4ad does https://github.com/nodejs/node/blob/main/doc/contributing/backporting-to-release-lines.md#how-to-submit-a-backport-pull-request help?

H4ad commented:
@RaisinTen Thanks, I'll try to backport by the end of this day.
